menpofit/dpm/fitter.py

```diff
@@ -0,0 +1,637 @@
+
+    Returns
+    -------
+    scores: `ndarray`
+        The (unary + pairwise) scores.
+    Ix: `dict`
+        Contains the coordinates of x for each part from the Generalised Distance Transform.
+    Iy: `dict`
+        Contains the coordinates of y for each part from the Generalised Distance Transform.
+    """
+    iy, ix = {}, {}
+    for depth in range(tree.maximum_depth, 0, -1):
+        for curr_vert in tree.vertices_at_depth(depth):
+            parent = tree.parent(curr_vert)
+            (ny, nx) = scores[parent].shape
+            w = def_coefs[curr_vert] * -1.00
+            (cx, cy, ds) = anchors[curr_vert]
+            step = 2**ds
+            virt_padx = (step - 1)*padding[0]
+            virt_pady = (step - 1)*padding[1]
+            startx = cx - virt_padx
+            starty = cy - virt_pady
+            msg, ix1, iy1 = call_shiftdt(scores[curr_vert], np.array(w, dtype=np.double), startx, starty, nx, ny, 1)
+            scores[parent] += msg
+            ix[curr_vert] = np.copy(ix1)
+            iy[curr_vert] = np.copy(iy1)
+    return scores, ix, iy
+
+
+def compute_filters_scores_only(scores, tree, filter_ids, ptr):
+    # might be needed when debugging
+    new_scores = []
+    for depth in range(0, tree.maximum_depth + 1):
+        for curr_vert in tree.vertices_at_depth(depth):
+            x, y = ptr[curr_vert]
+            curr_filter_id = filter_ids[curr_vert]
+            new_scores.append(scores[curr_filter_id][y, x])
+    return new_scores
+
+
+def _featpyramid(im, m_inter, sbin, pyra_pad):
+    # construct the featpyramid. For the time being, it has similar conventions as in
+    # Ramanan's code. TODO: In the future, use menpo's gaussian pyramid.
+    sc = 2 ** (1. / m_inter)
+    img_size = (im.shape[0], im.shape[1])
+    max_scale = int(log_m(min(img_size)*1./(5*sbin))/log_m(sc)) + 1
+    feats, scales = {}, {}
+    for i in range(m_inter):
+        sc_l = 1. / sc ** i
+        scaled = im.rescale(sc_l)
+        feats[i] = hog(scaled, mode='sparse', algorithm='zhuramanan', cell_size=sbin/2)
+        scales[i] = 2 * sc_l
+        feats[i + m_inter] = hog(scaled, mode='sparse', algorithm='zhuramanan', cell_size=sbin)
+        scales[i + m_inter] = sc_l
+
+        for j in range(i + m_inter, max_scale, m_inter):
+            scaled = scaled.rescale(0.5)
+            feats[j + m_inter] = hog(scaled, mode='sparse', algorithm='zhuramanan', cell_size=sbin)
+            scales[j + m_inter] = 0.5 * scales[j]
+
+    feats_np = {}  # final feats (keep only numpy array, get rid of the image)
+    for k, val in scales.iteritems():
+        scales[k] = sbin * 1. / val
+        feats_np[k] = np.pad(feats[k].pixels, ((0, 0), (pyra_pad[1] + 1, pyra_pad[1] + 1),
+                                               (pyra_pad[0] + 1, pyra_pad[0] + 1)), 'constant')
+
+    return feats_np, scales
+
+
+def _old_backtrack(x, y, tree, Ix, Iy, fsz, scale, pyra_pad):
```

Comment:
* ```_old_backtrack``` find the configurations for all the positions that score higher than the threshold in one go. 
* ```_backtrack``` does it one by one position because it also need to convert the configuration to be saved to **Qp**. 
* Notice that the function ```fast_fit_from_model``` use ```_old_backtrack``` since it does not concern about learning the model while the function ```fit_with_bb``` which is used in **DPMLearner** use ```_backtrack```. 
* Generally, if one does not concern training the model, ```_old_backtrack``` is faster than ```_backtrack```. 
* In future, I need to check that, for the function ```fit_with_bb```, if using ```_backtrack``` is actually faster than using ```_old_backtrack```  with ```_create_ex``` (comparing 1. iterate through a loop once doing all the work in each loop and  2. iterate though a loop twice and does half of the work in each loop)